### PR TITLE
Introduce in memory decision task state machine

### PR DESCRIPTION
**Pull request by andrewjdawson2016**
This diff does the following

- Introduces methods on mutable state to handle in memory decision tasks
- Consume these in memory decision tasks for query
- Update query state machine and query registry to not include expired state
- Fix the two places I missed including the mutability check from an earlier diff

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -18,6 +18,7 @@ require (
 	github.com/dgryski/go-farm v0.0.0-20190423205320-6a90982ecee2
 	github.com/eapache/go-resiliency v1.2.0 // indirect
 	github.com/emirpasic/gods v0.0.0-20190624094223-e689965507ab
+	github.com/facebookgo/clock v0.0.0-20150410010913-600d898af40a // indirect
```
Comment (wxing1292) [inline go.mod]: My guess, if you use go 1.13 this line will not be included

File: service/history/historyEngine.go [hunk 2]
```diff
@@ -585,44 +585,98 @@ func (e *historyEngineImpl) GetMutableState(
...
+	}
+	ttl := e.shard.GetConfig().LongPollExpirationInterval(domainCache.GetInfo().Name)
+	timer := time.NewTimer(ttl)
+	defer timer.Stop()
+
 	context, release, err := e.historyCache.getOrCreateWorkflowExecution(ctx, request.GetDomainUUID(), *request.GetExecution())
 	if err != nil {
 		return nil, err
 	}
-	ms, err := context.loadWorkflowExecution()
```
Comment (wxing1292) [inline service/history/historyEngine.go]: plz add a `defer func() { release(retError) }()` here
I know there is a release fn call after error checking and `msBuilder.GetQueryRegistry()`
the defer is more of a best practice

File: service/history/historyEngine.go [hunk 3]
```diff
@@ -585,44 +585,98 @@ func (e *historyEngineImpl) GetMutableState(
...
-		return nil, err
+
+	// after query is finished removed query from query registry and remove any potentially created in memory decision task
+	defer func() {
+		queryRegistry.removeQuery(queryID)
+		context, release, err := e.historyCache.getOrCreateWorkflowExecution(ctx, request.GetDomainUUID(), *request.GetExecution())
+		if err != nil {
+			retResp = nil
+			retErr = err
+		}
```
Comment (wxing1292) [inline service/history/historyEngine.go]: same here

File: service/history/historyEngine.go [hunk 4]
```diff
@@ -585,44 +585,98 @@ func (e *historyEngineImpl) GetMutableState(
...
+		release(nil)
+	}()
+
+	// ensure decision task exists to dispatch query on
+retryLoop:
+	for i := 0; i < conditionalRetryCount; i++ {
+		context, release, err = e.historyCache.getOrCreateWorkflowExecution(ctx, request.GetDomainUUID(), *request.GetExecution())
+		if err != nil {
+			return nil, err
+		}
```
Comment (wxing1292) [inline service/history/historyEngine.go]: same here

File: service/history/mutableState.go [hunk 5]
```diff
@@ -217,5 +217,12 @@ type (
 
 		CloseTransactionAsMutation(now time.Time, transactionPolicy transactionPolicy) (*persistence.WorkflowMutation, []*persistence.WorkflowEvents, error)
 		CloseTransactionAsSnapshot(now time.Time, transactionPolicy transactionPolicy) (*persistence.WorkflowSnapshot, []*persistence.WorkflowEvents, error)
+
+		AddInMemoryDecisionTaskScheduled(time.Duration) error
```
Comment (wxing1292) [inline service/history/mutableState.go]: could you put these APIs at very top?

File: service/history/mutableStateBuilder.go [hunk 6]
```diff
@@ -1505,12 +1505,20 @@ func (e *mutableStateBuilder) ReplicateWorkflowExecutionStartedEvent(
...
+	if err := e.checkMutability(opTag); err != nil {
+		return err
+	}
 	return e.decisionTaskManager.AddFirstDecisionTaskScheduled(startEvent)
 }
 
 func (e *mutableStateBuilder) AddDecisionTaskScheduledEvent(
 	bypassTaskGeneration bool,
 ) (*decisionInfo, error) {
+	opTag := tag.WorkflowActionDecisionTaskScheduled
```
Comment (wxing1292) [inline service/history/mutableStateBuilder.go]: nice

File: service/history/mutableStateBuilder.go [hunk 7]
```diff
@@ -3569,6 +3577,30 @@ func (e *mutableStateBuilder) CloseTransactionAsSnapshot(
 	return workflowSnapshot, workflowEventsSeq, nil
 }
 
+func (e *mutableStateBuilder) AddInMemoryDecisionTaskScheduled(ttl time.Duration) error {
```
Comment (wxing1292) [inline service/history/mutableStateBuilder.go]: could you plz put all business APIs before close transaction function?

File: service/history/mutableStateDecisionTaskManager_mock.go [hunk 8]
```diff
@@ -1,6 +1,6 @@
 // The MIT License (MIT)
 //
-// Copyright (c) 2019 Uber Technologies, Inc.
+// Copyright (c) 2017 Uber Technologies, Inc.
```
Comment (wxing1292) [inline service/history/mutableStateDecisionTaskManager_mock.go]: 2019

File: service/history/historyEngine.go [hunk 9]
```diff
@@ -585,44 +585,98 @@ func (e *historyEngineImpl) GetMutableState(
 func (e *historyEngineImpl) QueryWorkflow(
 	ctx ctx.Context,
 	request *h.QueryWorkflowRequest,
-) (*h.QueryWorkflowResponse, error) {
+) (retResp *h.QueryWorkflowResponse, retErr error) {
+	domainCache, err := e.shard.GetDomainCache().GetDomainByID(request.GetDomainUUID())
```
Comment (wxing1292) [inline service/history/historyEngine.go]: i would suggest 
1. moving this logic into a dedicated file.
2. breaking the 3 logic (loading mutable state and do something) into 3 functions
3. it is ok to have a for loop for retry, but plz move the logic within the for loop into a function

File: go.mod (hunk 1 above)
Comment (andrewjdawson2016) [reply go.mod]: thanks :)

File: service/history/historyEngine.go (hunk 2 above)
Comment (andrewjdawson2016) [reply service/history/historyEngine.go]: Oh thank you - I did it this way because I did not realize it was idempotent.

**Pull request merged**